### ISSUE TITLE
Make TS assume p5 is globally available

### ISSUE THREAD

**Issue by GoToLoop** [p5-types/p5.ts#80]
After we `import type p5 from "p5";` we can't do normal things such as:

- `new p5(sketch);`
- `const vel = p5.Vector.random2D();`
- `constructor(protected p = p5.instance) {}`

All those valid things above will make TS complain:
> 'p5' cannot be used as a value because it was imported using 'import type'.

However library "p5.js" always makes itself globally available as _p5_ after we import it inside an HTML file:
`<script defer src=https://cdn.JsDelivr.net/npm/p5></script>`

Therefore why doesn't "p5-types" recognize this fact? Is it a very long overdue oversight?

Anyways, I've ended up creating my own "p5-global.d.ts" type hint file as a workaround fix:
https://GitHub.com/GoSubRoutine/Steering-Text-Paths/blob/main/typings/p5-global.d.ts

After referencing that file as `/// <reference path="../typings/p5-global.d.ts" />`, 
it finally allows me to access library "p5.js" globally but w/ the following annoying caveat:

- `new window.p5(sketch);`
- `const vel = window.p5.Vector.random2D();`
- `constructor(protected p = window.p5.instance) {}`

I have to prefix _p5_ w/ `window.` everytime it's used as a value.

What I'm asking is for you devs to consider merging my workaround as just _p5_, w/o needing `window.`.

So every1 else can have the privilege to seamlessly use "p5.js" type hints on their sketches.

Besides allowing _p5_ to be used as a global value, "p5-global.d.ts" corrects _p5_ constructor's signature:
`constructor(sketch?: (p: p5) => void, node?: string | HTMLElement)`

Compared to current _p5_'s, both params are optional so we can empty invoke `new p5;` for global mode.

And it recognizes that 2nd param _node_ can accept a `string` as an alternative to an `HTMLElement`.

It also adds these 3 `static` properties to class _p5_:

1.  _VERSION_
2. _instance_
3. _disableFriendlyErrors_

Plus these 2 methods to class _p5_, which are very import when developing library addons for "p5.js" :

1. `registerMethod(register: 'init'|'pre'|'post'|'remove', callback: () => void): void`
2. `registerPreloadMethod(name: PropertyKey, proto: object): void`

Try out this example sketch repo inside an IDE so you can check how "p5-global.d.ts" works:

- https://GitHub.com/GoSubRoutine/Steering-Text-Paths
- https://GoSubRoutine.GitHub.io/Steering-Text-Paths/

**Comment by wuyudi** [comment]
![image](https://user-images.githubusercontent.com/30024051/212491467-d4880e45-25a0-4fd0-b51e-56f8b859a67e.png)

**Comment by GoToLoop** [comment]
After 3 weeks all I've got was a screenshot "answer" followed by "closed issue" action!

@wuyudi, the empty `new p5;` signature is a minor request in addition to the title's main issue:
# Make TS assume p5 is globally available

But given that's the only answer provided to me, let me refute it:

The empty `new p5;` constructor signature is very useful for global mode sketches when we need to prematurely access p5js API:
```js
new p5; // prematurely instantiate class p5. ;-)

const SQRT_3 = sqrt(3); // we can access p5's sqrt() method outside functions!

function setup() {
    print(SQRT_3);
}
```
Here's an online example sketch "Sierpinski Pentagon" which requires **cos()**, **sin()** and DEG_TO_RAD before callback **setup()**: https://OpenProcessing.org/sketch/955445

The other usage is to postpone p5js instantiation until we load assets from libraries which don't provide **preload()** loading.

In this Processing Discord topic, I've provided an example for pre-loading TensorFlow asset before p5js is started relying again on `new p5;`:
https://Discourse.Processing.org/t/problem-with-tensorflow-load-model/36390/4

In short, `constructor(sketch?: (p: p5) => void, node?: string | HTMLElement)` is the actual runtime signature for class p5!

Making the user-provided callback's argument _p_ optional is important for global mode sketches in 2 useful situations.

However the main issue is that the types provided by this repo ignore the fact that global variable _p5_ is available w/o any kind of `import`.

It's important that the type hints provided take into consideration that _p5_ is always available.

**Comment by wuyudi** [comment]
wow, I understand it, amazing!
so instead

```js
let a;
function setup(){
    a = sqrt(3);
}
```

we can write

```js
new p5;
const a = sqrt(3);
function setup(){
}
```

**Comment by wuyudi** [comment]
Would you mind mentioning a pull request so I can easily merge it?

**Comment by GoToLoop** [comment]
Sorry, I've got no idea how this "p5-types" repo auto-generates TS type annotations outta ["data.json"](https://raw.githubusercontent.com/processing/p5.js-website/main/src/templates/pages/reference/data.json). =(

B/c "data.json" isn't 100% accurate, some manual fixes are required to overwrite autogen.

Anyways, these are the changes I propose to be manually corrected, so "p5-types" will better represent the actual runtime types for p5*js library:

---
"index.d.ts":
-
```ts
declare global {
    p5: typeof p5
}

declare class p5 {
    //constructor(sketch: (...args: any[]) => any, node?: HTMLElement);
    constructor(sketch?: (p: object) => void, node?: string | HTMLElement);

    __proto__: typeof p5.prototype

    static VERSION: string
    static instance: p5

    static disableFriendlyErrors: boolean
    disableFriendlyErrors: boolean;

    canvas: HTMLCanvasElement

    _renderer: p5.Renderer
    _curElement: p5.Renderer

    registerMethod(
        register: 'init' | 'pre' | 'post' | 'remove',
        callback: (this: p5) => void
    ): void

    registerPreloadMethod(name: PropertyKey, proto: object): void
}
```
---
"src/core/structure.d.ts":
-
```ts
import * as p5 from '../../index';

declare module '../../index' {
    interface p5InstanceExtensions {
        //p5(sketch: object, node: string | object): void;
        p5: typeof p5
    }
}
```
---
"src/core/rendering.d.ts":
-
```ts
import * as p5 from '../../index';

declare module '../../index' {
    interface p5InstanceExtensions {
        //drawingContext: any;
        drawingContext:
            CanvasRenderingContext2D | 
            WebGLRenderingContext | 
            WebGL2RenderingContext;
    }
}
```
---
"src/core/environment.d.ts":
-
```ts
import * as p5 from '../../index';

declare module '../../index' {
    interface p5InstanceExtensions {
        //windowResized(event?: object): void;
        windowResized(event?: UIEvent): boolean | void;
    }
}
```
---
"src/events/keyboard.d.ts":
-
```ts
import * as p5 from '../../index';

declare module '../../index' {
    interface p5InstanceExtensions {
        //keyPressed(event?: object): void;
        keyPressed(event?: KeyboardEvent): boolean | void;

        //keyReleased(event?: object): void;
        keyReleased(event?: KeyboardEvent): boolean | void;

        //keyTyped(event?: object): void;
        keyTyped(event?: KeyboardEvent): boolean | void;
    }
}
```
---
"src/events/mouse.d.ts":
-
```ts
import * as p5 from '../../index';

declare module '../../index' {
    interface p5InstanceExtensions {
        //mouseMoved(event?: object): void;
        mouseMoved(event?: MouseEvent): boolean | void;

        //mouseDragged(event?: object): void;
        mouseDragged(event?: DragEvent): boolean | void;

        //mousePressed(event?: object): void;
        mousePressed(event?: MouseEvent): boolean | void;

        //mouseReleased(event?: object): void;
        mouseReleased(event?: MouseEvent): boolean | void;

        //mouseClicked(event?: object): void;
        mouseClicked(event?: MouseEvent): boolean | void;

        //doubleClicked(event?: object): void;
        doubleClicked(event?: MouseEvent): boolean | void;

        //mouseWheel(event?: object): void;
        mouseWheel(event?: WheelEvent): boolean | void;
    }
}
```
---
"src/events/touch.d.ts":
-
```ts
import * as p5 from '../../index';

declare module '../../index' {
    interface p5InstanceExtensions {
        //touchStarted(event?: object): void;
        touchStarted(event?: TouchEvent): boolean | void;

        //touchMoved(event?: object): void;
        touchMoved(event?: TouchEvent): boolean | void;

        //touchEnded(event?: object): void;
        touchEnded(event?: TouchEvent): boolean | void;

        //touches: object[];
        touches: {
            x: number,
            y: number,
            winX: number,
            winY: number,
            id: number
        }[];
    }
}
```
---

**Comment by wuyudi** [comment]
when I add

```ts
declare global {
    p5: typeof p5
}
```
, I meet https://stackoverflow.com/questions/69607719/statements-are-not-allowed-in-ambient-contexts , so I skip this.

Others should "work well". Though very dirty : (
